File: packages/quran-engine-dart/test/engine_test.dart
import 'dart:io';

import 'package:quran_engine/quran_engine.dart';
import 'package:test/test.dart';

/// Locate the repo data dir by walking up from the current directory to find
/// a `data/quran.json`.
String locateDataDir() {
  var dir = Directory.current;
  for (var i = 0; i < 12; i++) {
    final candidate = File(
        '${dir.path}${Platform.pathSeparator}data${Platform.pathSeparator}quran.json');
    if (candidate.existsSync()) {
      return '${dir.path}${Platform.pathSeparator}data';
    }
    final parent = dir.parent;
    if (parent.path == dir.path) break;
    dir = parent;
  }
  throw StateError('Could not locate data/quran.json');
}

void main() {
  late Engine engine;
  late Reciter alafasy;

  setUpAll(() async {
    engine = await Engine.load(dataDir: locateDataDir());
    // The Hafs "Mishary Alafasy" full-surah feed (afs) + ar.alafasy ayah feed.
    alafasy = engine.reciters.all().firstWhere((r) =>
        r.ayahIdentifier == 'ar.alafasy' &&
        r.surahLink == 'https://server8.mp3quran.net/afs/');
  });

  test('totalAyahs == 6236', () {
    expect(engine.quran.totalAyahs, 6236);
  });

  test('globalAyahNumber canonical cases', () {
    expect(engine.globalAyahNumber(1, 1), 1);
    expect(engine.globalAyahNumber(2, 1), 8);
    expect(engine.globalAyahNumber(114, 6), 6236);
  });

  test('surahAudioUrl(Alafasy, 1)', () {
    expect(surahAudioUrl(alafasy, 1),
        'https://server8.mp3quran.net/afs/001.mp3');
  });

  test('ayahAudioUrl(Alafasy, 8)', () {
    expect(ayahAudioUrl(alafasy, 8),
        'https://cdn.islamic.network/quran/audio/128/ar.alafasy/8.mp3');
  });

  test('juz boundaries', () {
    expect(engine.juz(1)!.startSurah, 1);
    expect(engine.juz(1)!.startAyah, 1);
    expect(engine.juz(30)!.endSurah, 114);
    expect(engine.juz(30)!.endAyah, 6);
  });

  test('juz from end + per-juz stats', () {
    final jp = engine.juzPage;
    expect(jp.juzFromEnd(1)!.id, 30);
    expect(jp.juzFromEnd(30)!.id, 1);
    expect(jp.juzFromEnd(0), isNull);
    expect(jp.juzFromEnd(31), isNull);

    final stats = jp.juzStats(30)!;
    expect(stats.ayahCount, jp.ayahsInJuz(30).length);
    expect(stats.surahCount >= 1 && stats.pageCount >= 1, isTrue);
    expect(stats.wordCount > 0 && stats.letterCount > 0, isTrue);
    expect(jp.juzStats(99), isNull);

    var sum = 0;
    for (var i = 1; i <= 30; i++) {
      sum += jp.juzStats(i)!.ayahCount;
    }
    expect(sum, 6236);
  });

  test('sortSurahs("ayahs","descending")[0].id == 2', () {
    final sorted = sortSurahs(engine.quran.all(), 'ayahs', 'descending');
    expect(sorted.first.id, 2);
  });

  test('parseReference("2:255")', () {
    final ref = engine.search.parseReference('2:255');
    expect(ref, isNotNull);
    expect(ref!.surah, 2);
    expect(ref.ayah, 255);
  });

  bool hits(List<VerseHit> r, int surah, int ayah) =>
      r.any((h) => h.surah == surah && h.ayah == ayah);

  test('regular search is pure substring (mid-word "orld" hits 1:2)', () {
    // "world" appears in the English translation of al-Fatihah 1:2
    // ("Lord of the worlds"); a mid-word substring must still match.
    final r = engine.search.searchVerses('orld');
    expect(hits(r, 1, 2), isTrue);
  });

  test('=lord (whole-word) hits 1:2 but =lor does not; plain lor does', () {
    expect(hits(engine.search.searchVerses('=lord'), 1, 2), isTrue);
    // whole-word: "lor" is not a complete token, so it must NOT match.
    expect(hits(engine.search.searchVerses('=lor'), 1, 2), isFalse);
    // plain substring: "lor" is inside "lord", so it DOES match.
    expect(hits(engine.search.searchVerses('lor'), 1, 2), isTrue);
  });

  test('"allah & 2" returns 0 (digit rejected before boolean path)', () {
    expect(engine.search.searchVerses('allah & 2'), isEmpty);
  });

  test('tajweed(1,1) span substrings match recorded text', () {
    final spans = engine.tajweedSpans(1, 1);
    expect(spans, isNotEmpty);
    final text = engine.ayah(1, 1)!.textArabic;
    for (final s in spans) {
      expect(s.text, text.substring(s.start, s.end));
    }
  });
}

File: packages/quran-engine-kotlin/src/test/kotlin/com/quranengine/EngineTest.kt
package com.quranengine

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestInstance
import java.io.File

/**
 * Canonical cases from `docs/PORTING.md`. The data dir is located by walking up from the working
 * directory to find `data/quran.json`.
 */
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class EngineTest {

    private lateinit var engine: Engine
    private lateinit var alafasy: Reciter

    @BeforeAll
    fun setup() {
        val dataDir = Engine.findDefaultDataDir()
        assertNotNull(dataDir, "Could not locate data/quran.json by walking up from the working dir")
        engine = Engine.load(dataDir)
        alafasy = engine.reciters.all().first { it.ayahIdentifier == "ar.alafasy" }
    }

    @Test
    fun totalAyahs() {
        assertEquals(6236, engine.quran.totalAyahs)
    }

    @Test
    fun globalAyahNumbers() {
        assertEquals(1, engine.quran.globalAyahNumber(1, 1))
        assertEquals(8, engine.quran.globalAyahNumber(2, 1))
        assertEquals(6236, engine.quran.globalAyahNumber(114, 6))
    }

    @Test
    fun surahAudio() {
        assertEquals("https://server8.mp3quran.net/afs/001.mp3", surahAudioUrl(alafasy, 1))
    }

    @Test
    fun ayahAudio() {
        assertEquals(
            "https://cdn.islamic.network/quran/audio/128/ar.alafasy/8.mp3",
            ayahAudioUrl(alafasy, 8),
        )
    }

    @Test
    fun juzBoundaries() {
        val j1 = engine.juzPage.juz(1)!!
        assertEquals(1, j1.startSurah)
        assertEquals(1, j1.startAyah)
        val j30 = engine.juzPage.juz(30)!!
        assertEquals(114, j30.endSurah)
        assertEquals(6, j30.endAyah)
    }

    @Test
    fun juzFromEndAndStats() {
        val jp = engine.juzPage
        assertEquals(30, jp.juzFromEnd(1)?.id)
        assertEquals(1, jp.juzFromEnd(30)?.id)
        assertNull(jp.juzFromEnd(0))
        assertNull(jp.juzFromEnd(31))

        val stats = jp.juzStats(30)!!
        assertEquals(jp.ayahsInJuz(30).size, stats.ayahCount)
        assertTrue(stats.surahCount >= 1 && stats.pageCount >= 1)
        assertTrue(stats.wordCount > 0 && stats.letterCount > 0)
        assertNull(jp.juzStats(99))

        val sum = (1..30).sumOf { jp.juzStats(it)!!.ayahCount }
        assertEquals(6236, sum)
    }

    @Test
    fun sorting() {
        val sorted = sortSurahs(engine.quran.all(), "ayahs", "descending")
        assertEquals(2, sorted[0].id) // Al-Baqarah, 286 ayahs
        assertEquals(286, sorted[0].numberOfAyahs)
    }

    @Test
    fun referenceParsing() {
        val ref = engine.search.parseReference("2:255")
        assertNotNull(ref)
        assertEquals(2, ref!!.surah)
        assertEquals(255, ref.ayah)
        assertEquals(AyahReference(2, 255), ref)
    }

    @Test
    fun tajweedSpansSubstringMatchesRecordedText() {
        val spans = engine.tajweed(1, 1)
        assertTrue(spans.isNotEmpty(), "expected tajweed spans for 1:1")
        val text = engine.quran.ayah(1, 1)!!.textArabic
        for (s in spans) {
            // The span's text must equal the exact UTF-16 slice of the ayah text.
            assertEquals(text.substring(s.start, s.end), s.text)
            assertNotNull(s.colorHex, "rule ${s.rule} should map to a color")
        }
        // Spot-check known annotations from data/tajweed/001.json:
        assertEquals("hamzatWaslSilent", spans[0].rule)
        assertEquals("#B4B4B4", spans[0].colorHex)
    }

    @Test
    fun verseSearchSubstringAndBooleanOperators() {
        val search = engine.search
        fun ids(results: List<Search.VerseIndexEntry>) = results.map { it.id }.toSet()

        // Regular (non-boolean) search is a PURE mid-word substring match: "orld" hits "worlds" in 1:2.
        assertTrue("1:2" in ids(search.searchVerses("orld")), "mid-word substring 'orld' should hit 1:2")

        // Whole-word operator `=`: `=lord` matches the token "lord" in 1:2.
        assertTrue("1:2" in ids(search.searchVerses("=lord")), "=lord should hit 1:2 (whole word)")

        // `=lor` is whole-word so it must NOT match (no token equals "lor")...
        assertTrue("1:2" !in ids(search.searchVerses("=lor")), "=lor should NOT hit 1:2 (not a whole word)")

        // ...but plain substring `lor` DOES match ("lord" contains "lor").
        assertTrue("1:2" in ids(search.searchVerses("lor")), "plain 'lor' substring should hit 1:2")

        // Digit rejection happens BEFORE the boolean branch: `allah & 2` returns no results.
        assertEquals(0, search.searchVerses("allah & 2").size, "a boolean query with a digit returns 0")
    }

    @Test
    fun dataDirContainsQuranJson() {
        val dir = Engine.findDefaultDataDir()!!
        assertTrue(File(dir, "quran.json").isFile)
    }
}
